sync/category.go

```diff
@@ -0,0 +1,80 @@
+package sync
+
+import (
+	"context"
+	"strconv"
+
+	"github.com/appuio/appuio-cloud-reporting/pkg/db"
+	"github.com/vshn/appuio-odoo-adapter/odoo/model"
+)
+
+// SyncCategory synchronizes model.InvoiceCategory in Odoo based on the given db.Category according to the following rules:
+//  * If db.Category.Target is NULL then it will convert and create a new model.InvoiceCategory and set db.Category.Target to the ID returned by Odoo.
+//  * If db.Category.Target has a value then it will search for a matching model.InvoiceCategory:
+//    * If not found, it will recreate the model.InvoiceCategory.
+//    * If found and model.InvoiceCategory.Name is up-to-date, it will return without error
+//    * If found and model.InvoiceCategory.Name differs from db.Category.Source, the model.InvoiceCategory is updated/reset.
+func (s *OdooSyncer) SyncCategory(ctx context.Context, category *db.Category) error {
+	if !category.Target.Valid {
+		return s.createCategoryInOdoo(ctx, category)
+	}
+	ic := model.InvoiceCategory{}
+	err := CategoryConverter{}.ToInvoiceCategory(category, &ic)
+	if err != nil {
+		return err
+	}
+
+	return s.updateCategoryIfNeeded(ctx, category, ic)
+}
+
+func (s *OdooSyncer) createCategoryInOdoo(ctx context.Context, category *db.Category) error {
+	ic := newInvoiceCategory(nil)
+	err := CategoryConverter{}.ToInvoiceCategory(category, &ic)
+	if err != nil {
+		return err
+	}
+	created, err := s.odoo.CreateInvoiceCategory(ctx, ic)
+	if err != nil {
+		return err
+	}
+	category.Target.String, category.Target.Valid = strconv.Itoa(created.ID), true
+	return nil
+}
+
+func (s *OdooSyncer) updateCategoryIfNeeded(ctx context.Context, category *db.Category, ic model.InvoiceCategory) error {
+	// TODO: re Idempotency, should we search by name instead? Maybe the category in odoo got deleted & recreated with a new ID manually
+	existingIC, err := s.odoo.FetchInvoiceCategoryByID(ctx, ic.ID)
+	if err != nil {
+		return err
+	}
+	if existingIC == nil {
+		// The category in Odoo might have been deleted, let's recreate
+		return s.createCategoryInOdoo(ctx, category)
+	}
+
+	cv := CategoryConverter{}
+	if !cv.IsSame(*existingIC, category) {
+		// Updating existing category should rarely be the case.
+		// Possible case is given if the category name has been manually updated in Odoo, in that case revert it since the DB is authoritative.
+		return s.updateCategoryInOdoo(ctx, category, existingIC)
+	}
```

Comment:
Should we also consider resetting the invoice category properties if they differ (e.g. subtotal, separator etc) ?

Reply:
Yes I would. The way I see it, the DB is the source of truth and categories should completely be managed by this operator. (Also the reason why I'm not a big fan of the adoption feature)